engine/application/public/input/action_manager.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+#include "common.hpp"
+#include "input_device_manager.hpp"
+#include <string>
+#include <variant>
+#include <vector>
+
+class InputDeviceManager;
+
+enum class DigitalActionType : uint8_t
+{
+    // Action happens once when input is received.
+    ePressed,
+    // Action happens continuously when input is being received.
+    eHold,
+    // Action happens once when input is released.
+    eReleased,
+};
+
+using DigitalInputAction = std::variant<GamepadButton, KeyboardCode, MouseButton>;
+using AnalogInputAction = GamepadAnalog;
+
+using DigitalInputActionList = std::vector<DigitalInputAction>;
+using AnalogInputActionList = std::vector<AnalogInputAction>;
+
+// Action for button inputs.
+struct DigitalAction
+{
+    std::string name {};
+    DigitalActionType type {};
+    DigitalInputActionList inputs {};
```

Comment:
input bindings is probably more appropriate for a name here, instead of input action (DigitalInputAction -> DigitalInputBinding, AnalogInputAction -> AnalogInputBinding).

Reply:
You're right, naming is hard. I'll change it